storage/innobase/srv/srv0start.cc

```diff
@@ -1988,6 +1985,22 @@ void innodb_preshutdown()
 
   if (srv_read_only_mode)
     return;
+
+#ifndef EMBEDDED_LIBRARY
+  if (srv_force_recovery < SRV_FORCE_NO_TRX_UNDO && srv_was_started &&
+      !trx_rollback_is_active && trx_sys.any_active_transaction_recovered())
+  {
+    /* Trx rollback thread may not be started yet, since error happened before
+       innobase_tc_log_recovery_done(), thus it is started here to finish the
+       normal shutdown process.
+     */
+    ib::warn() << "Start rollback task at preshutdown to rollback active "
+               << "transactions recovered.";
```

Comment:
If `innodb_fast_shutdown>1` we do not want to roll back any transactions. (Sorry, I did not realize this in my review yesterday.) We are missing a condition `srv_fast_shutdown <= 1 &&` here.

The use of `ib::logger` should be limited to those cases that require some special formatting, such as pretty-printing records. Here, we are unnecessarily invoking `operator<<` twice. It would be better to invoke the underlying function directly. I think that this should be `sql_print_information`, not `sql_print_warning`:
```c++
    sql_print_information("InnoDB: Rolling back recovered transactions at shutdown");
```
Even better, we could refactor the inner part of `trx_roll_report_progress()` to a new function (which would return `bool` indicating whether there were any transactions, and invoke it here:
```c++
  if (srv_force_recovery < SRV_FORCE_NO_TRX_UNDO && srv_was_started &&
      srv_fast_shutdown <= 1 && !trx_rollback_is_active &&
      trx_sys.report_to_rollback())
  {
    trx_rollback_is_active= true;
    srv_thread_pool->submit_task(&rollback_all_recovered_task);
  }
```
That common ‘inner’ function `trx_sys_t::report_to_rollback()` would count transactions and rows:
```c++
  // … declare arg and iterate over `rw_trx_hash`
  if (!arg.n_trx)
    return false;
  service_manager_extend_timeout(INNODB_EXTEND_TIMEOUT_INTERVAL,
		 		 "To roll back: " UINT32PF " transactions, "
				 UINT64PF " rows", arg.n_trx, arg.n_rows);
  sql_print_information("InnoDB: To roll back: " UINT32PF " transactions, "
                        UINT64PF " rows", arg.n_trx, arg.n_rows);
  return true;
```
In this way, we would have a consistent type of output.

Reply:
Hi Marko,  disable_log_notes is set to 1 at unireg_abort(), information logs will be dropped. That is why I use warning here. The warning happens only when error happens. I run the whole test. It doesn't break other tests.

Reply:
> If innodb_fast_shutdown>1 we do not want to roll back any transactions. (Sorry, I did not realize this in my review yesterday.) We are missing a condition srv_fast_shutdown <= 1 && here.

Looks the code in `logs_empty_and_mark_files_at_shutdown()`, it will wait for all active transactions to either rollback or deregistered.  It seams that is not expected according to the comment of srv_fast_shutdown. 
If you want it to be fixed in this patch. I think it should be fixed in trx_rollback_recovered() like:
```
--- a/storage/innobase/trx/trx0roll.cc
+++ b/storage/innobase/trx/trx0roll.cc
@@ -749,8 +749,8 @@ void trx_rollback_recovered(bool all)
     ut_ad(trx_state_eq(trx, TRX_STATE_ACTIVE));
     ut_d(trx->mutex_unlock());
 
-    if (srv_shutdown_state != SRV_SHUTDOWN_NONE && !srv_undo_sources &&
-        srv_fast_shutdown)
+    if (srv_shutdown_state != SRV_SHUTDOWN_NONE &&
+        (srv_fast_shutdown > 1 || !srv_undo_sources))
       goto discard;
```

Reply:
Thank you. I did not know about `disable_log_notes`. The pre-shutdown was written by @vaintroub in 5e62b6a5e06eb02cbde1e34e95e26f42d87fce02. It feels wrong to have all informational messages suppressed during shutdown. Occasionally there are shutdown hangs, and it would be useful to know based on logs, at which exact point the shutdown is hanging.

Regarding `srv_fast_shutdown`, I meant that there is no point in scheduling a task if we do not expect it to do anything useful. The purpose of `innodb_fast_shutdown=2` is to execute a crash-like shutdown (not writing out dirty pages and updating the log checkpoint). The purpose of `innodb_fast_shutdown=3` is to execute a clean shutdown, but without rolling back any recovered transactions. That condition is being checked at a deeper level, in `row_undo_step()`:
```c++
	if (UNIV_UNLIKELY(!trx->dict_operation
			  && !srv_undo_sources
			  && srv_shutdown_state != SRV_SHUTDOWN_NONE)
	    && (srv_fast_shutdown == 3 || trx == trx_roll_crash_recv_trx)) {
		/* Shutdown has been initiated. */
		trx->error_state = DB_INTERRUPTED;
		return NULL;
	}
```
I just realized that for recovered transactions (`trx == trx_roll_crash_recv_trx`) this does not even depend on `innodb_fast_shutdown=3`. We will always interrupt the rollback of any recovered non-DDL transactions on shutdown. If `innodb_fast_shutdown=3`, we will additionally interrupt the rollback of any active transaction. Typically, shutdown would cause a rollback of all currently active transactions, and normally shutdown would wait until they have been rolled back.

I suspect that sometimes, small transactions may be rolled back nevertheless, because the `srv_shutdown_state` would be changed later. This should be tested with a larger transaction, such as a `DELETE` or `UPDATE` of a large number of rows.

Reply:
You are right, so we don't need to change the code in trx_rollback_recovered, but we still need to start the rollback task.

Reply:
If there is anything I suppressed in pre-shutdown, it would be by oversight. I certainly do not remember doing anything with messages. The reason it exist is to run purge to the end, in slow shutdown. purge might need some audit plugin, if I remember correctly, so there needs to be a time during shutdown, when users are gone, but plugins are still intact. 

Before me, (and before tpool) the sequencing was done by some innodb_thd_proxy.